apps/studio/src/server/modules/page/page.router.ts

```diff
@@ -30,11 +36,112 @@ import { getSiteConfig, getSiteNameAndCodeBuildId } from "../site/site.service"
+    ? // NOTE: If this is at root, we will always use `null` as the parent
+      // otherwise, this is a `create` action and the parent of the resource that
+      // we want to create is the resource passed in.
+      // However, because we don't have root level permissions for now,
+      // we will pass in `null` to signify the site level permissions
+      { parentId: null }
+    : await db
+        .selectFrom("Resource")
+        .where("Resource.id", "=", resourceId)
+        .select(["Resource.parentId"])
+        .executeTakeFirstOrThrow()
+
+  const perms = await definePermissionsFor({ ...rest, resourceId: null })
+
+  // TODO: create should check against the current resource id
+  if (perms.cannot(action, resource)) {
+    throw new TRPCError({
+      code: "FORBIDDEN",
+      message: "You do not have sufficient permissions to perform this action",
+    })
+  }
+}
+
+// TODO: Need to do validation like checking for existence of the page
+// and whether the user has write-access to said page: replace protectorProcedure in this with the new procedure
+const validatedPageProcedure = protectedProcedure.use(
+  async ({ next, rawInput }) => {
+    if (
+      typeof rawInput === "object" &&
+      rawInput !== null &&
+      "content" in rawInput
+    ) {
+      // NOTE: content will be the entire page schema for now...
+      if (!schemaValidator(safeJsonParse(rawInput.content as string))) {
+        throw new TRPCError({
+          code: "BAD_REQUEST",
+          message: "Schema validation failed.",
+          cause: schemaValidator.errors,
+        })
+      }
+    } else {
+      throw new TRPCError({
+        code: "BAD_REQUEST",
+        message: "Missing request parameters.",
+      })
+    }
+
+    return next()
+  },
+)
+
 export const pageRouter = router({
+  list: protectedProcedure
+    .input(
+      z.object({
+        siteId: z.number(),
+        resourceId: z.number().optional(),
+      }),
+    )
+    .query(async ({ ctx, input: { siteId, resourceId } }) => {
+      await validateUserPermissions({
+        userId: ctx.user.id,
+        siteId,
+        action: "read",
+      })
+
+      let query = db
+        .selectFrom("Resource")
+        .where("Resource.siteId", "=", siteId)
+
+      if (resourceId) {
+        query = query.where("Resource.parentId", "=", String(resourceId))
+      }
+      return query
+        .select([
+          "Resource.id",
+          "Resource.permalink",
+          "Resource.title",
+          "Resource.publishedVersionId",
+          "Resource.draftBlobId",
+          "Resource.type",
+        ])
+        .execute()
+    }),
+
   readPage: protectedProcedure
     .input(basePageSchema)
-    .output(readPageOutputSchema)
-    .query(async ({ input: { pageId, siteId } }) => {
+    .query(async ({ ctx, input: { pageId, siteId } }) => {
+      await validateUserPermissions({
+        userId: ctx.user.id,
+        siteId,
+        action: "read",
+      })
```

Comment:
thought(non-blocking): actually why do we not want to just pass in the pageId/resourceId here, and leave it to just `validateUserPermissions` to determine if it wants to make use of the information?

Not a big deal tbh so don't need to make changes at least for launch, just maybe wondering if I should adjust my thought process a little bit here regarding the permissions model.